Block/Catalog/Product/View.php

```diff
@@ -90,84 +109,114 @@ public function __construct(
         WidgetConfigHelper $widgetConfigHelper,
         Config $config,
         Resolver $localeResolver,
+        Session $catalogSession,
+        Logger $logger,
+        ProductRepository $productRepository,
         array $data = []
     ) {
         parent::__construct($context, $data);
+        $this->logger = $logger;
+        $this->catalogSession = $catalogSession;
         $this->config = $config;
         $this->registry = $registry;
         $this->localeResolver = $localeResolver;
-        $this->getProduct();
-        $this->getPlans();
         $this->apiConfigHelper = $apiConfigHelper;
         $this->widgetConfigHelper = $widgetConfigHelper;
+        $this->productRepository = $productRepository;
     }
 
     /**
-     * @return void
-     * @throws LocalizedException
+     * Get last viewed product id session
+     * @return int
      */
-    private function getProduct()
+    private function getLastViewProductId(): int
     {
-        $this->product = $this->registry->registry('product');
-        if (!$this->product->getId()) {
-            throw new LocalizedException(__('Failed to initialize product'));
+        $this->logger->info('$this->catalogSession->getData()', [$this->catalogSession->getData()]);
+        return intval($this->catalogSession->getData('last_viewed_product_id'));
+    }
+
+    /**
+     * @return ProductInterface
+     * @throws AlmaProductException
+     */
+    private function getProduct(): ProductInterface
+    {
+        if (!is_null($this->product)) {
+            return $this->product;
+        }
+
+        try {
+            $this->product = $this->productRepository->getById($this->getLastViewProductId());
+        } catch (NoSuchEntityException $e) {
+            $this->logger->error('No product in product Repository with this id', ['Exception message' => $e->getMessage(), 'Product ID' => $this->getLastViewProductId()]);
+            throw new AlmaProductException(sprintf('No product in product Repository with this id %s', $this->getLastViewProductId()));
+        }
+
+        if (is_null($this->product)) {
+            $this->product = $this->registry->registry('product');
+            $this->logger->info('with registry', [$this->product]);
+        }
+        if (is_null($this->product)) {
```

Comment:
tu peux mettre ce if dans le if d'au dessus

Reply:
Ca va rajouter de la complexité d'imbriquer les if non ?

Reply:
la tu vas faire un test pour rien dans le cas ou le produit n'est plus nul

Reply:
@ClaireBonacheraAlma I am agree with you (after many reflexion on how to calculate the cognitive complexity of imbricated `if` 😉 )

@Francois-Gomis , I thank the same as you (`if` imbrication is more complex to read), but I realise that Claire is right, the second `is_null` will always be evaluated, even if `$this->productRepository->getById()` exit with success... So I change my mind, I think the complexity of imbricated `if` in this method is less important than the performance given by one less `if is_null`